### PR TITLE
Tests | Added more datatype tests to increase code coverage

### PR DESCRIPTION
**Pull request by lilgreenbird**

### CODE REVIEW

File: src/test/java/com/microsoft/sqlserver/jdbc/TestResource.java [hunk 1]
```diff
@@ -153,6 +153,7 @@ public static String getResource(String key) {
             {"R_expectedValue", "Expected value: "}, {"R_expectedValueAtIndex", "Expected value at index: "},
             {"R_switchFailed", "Switch case is not matched with data"},
             {"R_resultsetNotInstance", "Result set is not instance of SQLServerResultSet"},
-
+            {"R_resultsetNotInstance", "Result set is not instance of SQLServerResultSet"},
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestResource.java]: Duplicate Entry?

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        String connectionString = TestUtils.getConfiguredProperty("mssql_jdbc_test_connection_properties");
+        String tableName = RandomUtil.getIdentifier("TFOO3");
+        String procName = RandomUtil.getIdentifier("SPFOO3");
+
+        try (Connection conn = DriverManager.getConnection(connectionString)) {
+            try (Statement stmt = conn.createStatement()) {
+                try {
+                    stmt.executeUpdate("DROP TABLE " + AbstractSQLGenerator.escapeIdentifier(tableName));
+                    stmt.executeUpdate(" DROP PROCEDURE " + AbstractSQLGenerator.escapeIdentifier(procName));
+                } catch (Exception e) {}
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java]: Exception should not fall through if the test fails to drop these tables. Test should abort with an appropriate error message.

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                cstmt.registerOutParameter((int) 5, (int) 5);
+                cstmt.setObject((int) 4, Short.valueOf("-5372"), (int) 5);
+
+                // get results and a value
+                ResultSet rs = cstmt.executeQuery();
+                rs.next();
+
+                if (rs.getInt(1) != 0) {
+                    fail("Received data not equal to setdata");
+
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java]: Bunch of unnecessary spaces here and below.

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java [hunk 4]
```diff
@@ -623,6 +625,8 @@ public static Object roundSmallDateTimeValue(Object value) {
      * @return boolean
      */
     public static boolean serverSupportsDataClassification(Statement stmt) {
+
```
Comment (peterbae) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java]: remove

File: src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java [hunk 5]
```diff
@@ -33,9 +34,14 @@
 
 import com.microsoft.sqlserver.jdbc.ISQLServerResultSet;
 import com.microsoft.sqlserver.jdbc.RandomUtil;
+import com.microsoft.sqlserver.jdbc.SQLServerResultSet;
 import com.microsoft.sqlserver.jdbc.TestUtils;
 import com.microsoft.sqlserver.testframework.AbstractSQLGenerator;
 import com.microsoft.sqlserver.testframework.AbstractTest;
+import com.microsoft.sqlserver.testframework.DBConnection;
```
Comment (peterbae) [inline src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: do we need to import these?

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    /*
+     * I have created two files with wrong nano value (-1) and wrong offset (15*60) by "editing" couple of generated
+     * serialized files I have checked them in here so they can be used verify the error handling. Note the code to
+     * generate this is checked in as a text file in the same dir. DateTimeOffset.java.txt
+     */
+    // Negative test cases.
+    private static void verifyMessedSerialization() throws Exception {
+        // these values are from the wrongnanos.dat wrongoffset.dat files checked in with the code
+        // the values represent the serialized DTO class with invalid values.
+        byte wrongnanos[] = {-84, -19, 0, 5, 115, 114, 0, 47, 109, 105, 99, 114, 111, 115, 111, 102, 116, 46, 115, 113,
```
[the remark this answers is not in the document]
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java]: this made me realize I missed the reference file.
The names should match.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DateTimeOffset.java.txt [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+    private void readObject(java.io.ObjectInputStream stream) throws java.io.InvalidObjectException
+    {
+        // For added security/robustness, the only way to rehydrate a serialized DateTimeOffset
+        // is to use a SerializationProxy.  Direct use of readObject() is not supported.
+        //
+        // Note that the reason provided for the InvalidObjectException is deliberately less
+        // than specific.  Anyone trying to "hack" this class doesn't need to know why we
+        // didn't let them...
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DateTimeOffset.java.txt]: Should this comment be ported to a public repo?

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DateTimeOffset.java.txt [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+		File simple = new File("d:\\temp\\wrongoffset.dat");
+		// 15 hours offset not valid
+		DateTimeOffset dtm = new DateTimeOffset(1, 300, 15*60);
+		createBadSerializedForm(simple, dtm);
+	}
+	static void createBadSerializedForm(File simple, DateTimeOffset dtm) throws Exception
+	{
+		FileOutputStream so = new java.io.FileOutputStream(simple);
+	    java.io.ObjectOutput s = new java.io.ObjectOutputStream(so);
+	   	s.writeObject(dtm);
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DateTimeOffset.java.txt]: Weird spacing

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DateTimeOffset.java.txt [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+ */
+
+
+/**
+ *  			!DUMMY CLASS!
+ *  
+ *  
+ */
+public final class DateTimeOffset extends Object implements java.io.Serializable
+{
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DateTimeOffset.java.txt]: Please format this file, the bracketing is off.

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java [hunk 10]
```diff
@@ -666,4 +668,34 @@ public static boolean supportJDBC43(Connection con) throws SQLException {
...
+     * @param con
+     *        connection to server
+     * @return boolean
+     * @throws SQLException
+     */
+    public static boolean isSqlAzure(Connection con) throws SQLException {
+        try (ResultSet rs = con.createStatement().executeQuery("SELECT CAST(SERVERPROPERTY('EngineEdition') as INT)")) {
+            rs.next();
+            int engineEdition = rs.getInt(1);
+            return (engineEdition == ENGINE_EDITION_FOR_SQL_AZURE || engineEdition == ENGINE_EDITION_FOR_SQL_AZURE_DW);
```
Comment (peterbae) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java]: Can you use a static variable (like how it's done in our fx framework) to mark if the server's type has been determined and re-use that, so that we don't have to contact the server more than once?

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+import static org.junit.jupiter.api.Assertions.fail;
+
+
+@RunWith(JUnitPlatform.class)
+public class CallableMixedTest {
+
+    @Test
+    public void datatypestest() throws Exception {
+        Class.forName("com.microsoft.sqlserver.jdbc.SQLServerDriver");
+        String connectionString = TestUtils.getConfiguredProperty("mssql_jdbc_test_connection_properties");
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java]: Please Extend this class with AbstractTest class that already reads this connection property.

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+import org.junit.runner.RunWith;
+import static org.junit.jupiter.api.Assertions.fail;
+
+
+@RunWith(JUnitPlatform.class)
+public class CallableMixedTest {
+
+    @Test
+    public void datatypestest() throws Exception {
+        Class.forName("com.microsoft.sqlserver.jdbc.SQLServerDriver");
```
Comment (peterbae) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java]: I dont think we need these lines to load that class into classpath for our tests

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/BigIntegerTest.java [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+                testSetObject(AbstractSQLGenerator.escapeIdentifier(tableName), BigInteger.valueOf(Long.MAX_VALUE),
+                        row++, pstmt, TestType.SETOBJECT_WITHTYPE);
+
+                testSetObject(AbstractSQLGenerator.escapeIdentifier(tableName), BigInteger.valueOf(Long.MIN_VALUE),
+                        row++, pstmt, TestType.SETOBJECT_WITHTYPE);
+                testSetObject(AbstractSQLGenerator.escapeIdentifier(tableName), BigInteger.valueOf(10), row++, pstmt,
+                        TestType.SETOBJECT_WITHTYPE);
+                testSetObject(AbstractSQLGenerator.escapeIdentifier(tableName), BigInteger.valueOf(-10), row++, pstmt,
+                        TestType.SETOBJECT_WITHTYPE);
+                testSetObject(AbstractSQLGenerator.escapeIdentifier(tableName), BigInteger.ZERO, row++, pstmt,
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/BigIntegerTest.java]: Instead of multiple calls to `AbstractSQLGenerator.escapeIdentifier(tableName)` maintain `escapedTableName` instead?

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/BigIntegerTest.java [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+                        TestType.SETOBJECT_WITHOUTTYPE);
+                testSetObject(AbstractSQLGenerator.escapeIdentifier(tableName), bigIntNeg, row++, pstmt,
+                        TestType.SETOBJECT_WITHOUTTYPE);
+
+                // Test setNull
+                testSetObject(AbstractSQLGenerator.escapeIdentifier(tableName), bigIntNeg, row++, pstmt,
+                        TestType.SETNULL);
+
+                try {
+                    stmt.executeUpdate("drop table " + AbstractSQLGenerator.escapeIdentifier(tableName));
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/BigIntegerTest.java]: Use TestUtils > dropTableIfExists for consistency?

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/KatmaiDataTypesTest.java [hunk 15]
```diff
@@ -0,0 +1,1684 @@
...
+                }
+            }
+        }
+
+        /*
+         * For testing the setObject and setNull methods in PreparedStatement, use the verifySetter* methods. These
+         * methods prepare a single statement and execute it for all different data types by calling the appropriate
+         * 'setObject' methods for each data type and/or type conversion.
+         */
+        abstract void verifySetters(PreparedStatement ps) throws Exception;
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/KatmaiDataTypesTest.java]: Please move all abstract methods to the top of the class with Javadocs instead of comment.

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java [hunk 16]
```diff
@@ -666,4 +673,38 @@ public static boolean supportJDBC43(Connection con) throws SQLException {
...
+    
+    /**
+     * Returns if connected to SQL Azure
+     * @param con
+     *        connection to server
+     * @return boolean
+     * @throws SQLException
+     */
+    public static boolean isSqlAzure(Connection con) throws SQLException {
+        if (_determinedSqlAzureOrSqlServer) {
```
Comment (peterbae) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java]: where are we setting _determinedSqlAzureOrSqlServer = true? also, this needs to be applied to isSqlAzureDW as well.

Comment (peterbae) [reply src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java]: also you need to set _isSqlAzure

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/SparseTest.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+                bd.append("create table " + escapedTableName + " (col1 int, col2 varbinary(max)");
+                for (int i = 3; i <= 1024; i++) {
+                    bd.append(", col" + i + " varchar(20) SPARSE NULL");
+                }
+                bd.append(")");
+                String query = bd.toString();
+
+                stmt.executeUpdate(query);
+
+                stmt.executeUpdate("insert into " + escapedTableName + " (col1, col2, col1023)values(1, 0x45, 'yo')");
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/SparseTest.java]: This test doesn't really make a lot of sense. Why did it create 1024 columns if it's just going to use one of them and leave the rest as "NULL". It also only verifies with col1023, and not col1 or col2. Can we clean this up a bit.

File: src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java [hunk 18]
```diff
@@ -329,4 +329,141 @@ public void testGetterOnNull() throws SQLException {
...
+
+            int connHold = con.getHoldability();
+            assertEquals(stmt.getResultSetHoldability(), connHold);
+            assertEquals(rs.getHoldability(), connHold);
+
+            for (int i = 0; i < holdabilityOptions.length; i++) {
+
+                if ((connHold = con.getHoldability()) != holdabilityOptions[i]) {
+                    con.setHoldability(holdabilityOptions[i]);
+                    assertEquals(con.getHoldability(), holdabilityOptions[i]);
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: The purpose of setting holdability on a connection is so that it persists when retrieving result sets from said connection. Verifying setting/getting holdability on a connection object is a test with a pretty low chance of failure. Can we also create a ResultSet from the Connection here and verify the connection's holdability did in fact persist? Even better if we actually add scenarios where holdability behaves as expected instead of just checking getters. The latter can probably be done as an improvement though.

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/SparseTest.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
+package com.microsoft.sqlserver.jdbc.datatypes;
+
+import java.sql.*;
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/SparseTest.java]: Replace with only necessary imports.

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java [hunk 20]
```diff
@@ -0,0 +1,102 @@
+package com.microsoft.sqlserver.jdbc.callablestatement;
+
+import java.sql.*;
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java]: Replace with only necessary imports.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java [hunk 21]
```diff
@@ -0,0 +1,176 @@
+package com.microsoft.sqlserver.jdbc.unit.serial;
+
+import java.sql.*;
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java]: Replace with only necessary imports.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java [hunk 22]
```diff
@@ -0,0 +1,176 @@
+package com.microsoft.sqlserver.jdbc.unit.serial;
+
+import java.sql.*;
+
+import org.junit.jupiter.api.Test;
+import org.junit.platform.runner.JUnitPlatform;
+import org.junit.runner.RunWith;
+
+import com.microsoft.sqlserver.jdbc.*;
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java]: Replace with only necessary imports.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.platform.runner.JUnitPlatform;
+import org.junit.runner.RunWith;
+
+import com.microsoft.sqlserver.jdbc.*;
+import com.microsoft.sqlserver.testframework.AbstractTest;
+
+import microsoft.sql.*;
+
+import java.io.*;
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java]: Replace with only necessary imports.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+
+        exThrown = false;
+        try {
+            verifyMessedSerializationHelper(wrongOffset);
+        } catch (IllegalArgumentException e) {
+            exThrown = true;
+        }
+
+        if (!exThrown) {
+            fail("wrongnanos serialized form succeeded.");
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java]: wrong nanos, the message is also not very useful.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java [hunk 25]
```diff
@@ -1111,20 +1111,20 @@ public void testJdbc41CallableStatementMethods() throws Exception {
                                 + "(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}")) {
                     cstmt.registerOutParameter(1, java.sql.Types.VARCHAR);
                     cstmt.registerOutParameter(2, java.sql.Types.INTEGER);
-                    cstmt.registerOutParameter(3, java.sql.Types.FLOAT);
-                    cstmt.registerOutParameter(4, java.sql.Types.DECIMAL);
+                    cstmt.registerOutParameter("col3Value", java.sql.Types.FLOAT);
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java]: Why are these changes necessary?

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+import microsoft.sql.*;
+
+import java.io.*;
+
+import static org.junit.jupiter.api.Assertions.fail;
+
+
+@RunWith(JUnitPlatform.class)
+public class DTOSerialTest extends AbstractTest {
+    private static final String dateString = "2007-05-08 12:35:29.1234567 +12:15";
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java]: Hard-coded value, so this will only ever test with this specific dateString? Might be a little hard to refactor considering the serialization test below but this probably isn't a good way to be testing these APIs.

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/KatmaiDataTypesTest.java [hunk 27]
```diff
@@ -0,0 +1,1673 @@
...
+
+import java.sql.*;
+import java.text.MessageFormat;
+
+import com.microsoft.sqlserver.jdbc.*;
+import com.microsoft.sqlserver.testframework.AbstractSQLGenerator;
+import com.microsoft.sqlserver.testframework.AbstractTest;
+
+import java.math.*;
+import java.util.*;
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/KatmaiDataTypesTest.java]: Replace with only necessary imports.

File: src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java (hunk 18 above)
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: that is tested elsewhere this test just tests getter/setter

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java (hunk 25 above)
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java]: code coverage :)
this always calls the same registerOutParameter it's good to test the others as well

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java [hunk 28]
```diff
@@ -64,8 +64,16 @@
...
+    
     // private static SqlType types = null;
     private static ArrayList<SqlType> types = null;
+    private final static int ENGINE_EDITION_FOR_SQL_AZURE = 5;
+    private final static int ENGINE_EDITION_FOR_SQL_AZURE_DW = 6;
+
+    // whether we determined if the target server is SQL Azure or DW
+    private static boolean _determinedSqlAzureOrSqlServer = false;
+    private static boolean _isSqlAzure = false;
+    private static boolean _isSqlAzureDW = false;
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java]: Let's use the same methods to check for Azure DB/DW as in PR #903, implemented in AbstractTest class.

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+                    if (cstmt.getInt((int) 1) != -2147483648) {
+                        fail("Received data not equal to setdata");
+                    }
+
+                    if (cstmt.getInt((int) 5) != -5372) {
+                        fail("Received data not equal to setdata");
+                    }
+                }
+
+                try (ResultSet rs = cstmt.executeQuery()) {}
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java]: Let's remove this call, if not required to be tested..

File: src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java [hunk 30]
```diff
@@ -329,4 +329,145 @@ public void testGetterOnNull() throws SQLException {
...
+                do {
+                    rs.moveToCurrentRow();
+                    rs.deleteRow();
+                    assert (rs.rowDeleted());
+                } while (rs.next());
+
+            } catch (Exception e) {
+                fail(e.toString());
+            } finally {
+                stmt.executeUpdate("drop table " + AbstractSQLGenerator.escapeIdentifier(tableName));
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/resultset/ResultSetTest.java]: Use TestUtil functions

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java [hunk 31]
```diff
@@ -0,0 +1,199 @@
...
+            try {
+                stmt.executeUpdate("RAISERROR ('foo', 13,1) WITH LOG");
+
+            } catch (SQLServerException x) {
+                currException = x;
+            }
+
+            // store the info
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            ObjectOutputStream out = new ObjectOutputStream(bos);
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java]: Streams should be created in try-with-resources block so they are closed in the end.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java [hunk 32]
```diff
@@ -0,0 +1,199 @@
...
+        }
+
+        if (!exThrown) {
+            fail("serialized form with wrong offset unexpectedly succeeded");
+        }
+    }
+
+    private static void verifyMessedSerializationHelper(byte[] svalue) throws Exception {
+        ObjectInputStream in = new ObjectInputStream(new ByteArrayInputStream(svalue));
+        DateTimeOffset dtn = (DateTimeOffset) in.readObject();
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java]: What is the purpose reading from these Streams? Should validate `dtn`?

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java [hunk 33]
```diff
@@ -1111,20 +1111,20 @@ public void testJdbc41CallableStatementMethods() throws Exception {
...
+                    cstmt.registerOutParameter(4, java.sql.Types.DECIMAL, 2);
                     cstmt.registerOutParameter(5, microsoft.sql.Types.GUID);
                     cstmt.registerOutParameter(6, java.sql.Types.SQLXML);
-                    cstmt.registerOutParameter(7, java.sql.Types.VARBINARY);
+                    cstmt.registerOutParameter("col7Value", java.sql.Types.VARBINARY, 10);
```
Comment (cheenamalhotra) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java]: Reason for changes in this test?

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java (hunk 28 above)
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java]: manually merged

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java (hunk 33 above)
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/StatementTest.java]: for coverage, calls other registerOutParameter with different args

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java (hunk 28 above)
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java]: manually merged for now

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java (hunk 32 above)
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/unit/serial/DTOSerialTest.java]: added comment, this will throw error for negative tests

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/BigIntegerTest.java [hunk 34]
```diff
@@ -0,0 +1,232 @@
...
+        SETOBJECT_WITHTYPE, // This is to test conversions in Table B-5
+        SETOBJECT_WITHOUTTYPE, // This is to test conversions in Table B-4
+        SETNULL // This is to test setNull method
+    };
+
+    final static String tableName = RandomUtil.getIdentifier("BigIntegerTestTable");
+    final static String escapedTableName = AbstractSQLGenerator.escapeIdentifier(tableName);
+
+    @Test
+    public void testJDBC41BigInteger() throws Exception {
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/BigIntegerTest.java]: Please remove `JDBC41` from the test name

File: src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java [hunk 35]
```diff
@@ -0,0 +1,95 @@
...
+
+                // get the param without getting the resultset
+                try (ResultSet rs = cstmt.executeQuery()) {
+                    assertEquals(cstmt.getInt((int) 1), -2147483648);
+                }
+
+                try (ResultSet rs = cstmt.executeQuery()) {
+                    rs.next();
+
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/callablestatement/CallableMixedTest.java]: Weird spacing between lines in this test. Please cleanup.

File: src/test/java/com/microsoft/sqlserver/jdbc/datatypes/BigIntegerTest.java [hunk 36]
```diff
@@ -0,0 +1,232 @@
...
+                    // Get the data first before calling rs.wasNull()
+                    rs.getString(i);
+                    assertEquals(rs.wasNull(), true);
+                }
+                return;
+            }
+
+            if ((0 > obj.compareTo(BigInteger.valueOf(Long.MIN_VALUE)))
+                    || (0 < obj.compareTo(BigInteger.valueOf(Long.MAX_VALUE)))) {
+                // For the BigInteger values greater/less than Long limits test only the long data type.
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/datatypes/BigIntegerTest.java]: Please use multi-line comments. This applies to all tests.
